services/filesstore/s3store.go

```diff
@@ -73,20 +78,36 @@ func (b *S3FileBackend) TestConnection() *model.AppError {
 		return model.NewAppError("TestFileConnection", "api.file.test_connection.s3.connection.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
-	exists, err := s3Clnt.BucketExists(context.Background(), b.bucket)
-	if err != nil {
-		return model.NewAppError("TestFileConnection", "api.file.test_connection.s3.bucket_exists.app_error", nil, err.Error(), http.StatusInternalServerError)
+	exists := true
+	// If a path prefix is present, we attempt to test the bucket by listing objects under the path
+	// and just checking the first response. This is because the BucketExists call is only at a bucket level
+	// and sometimes the user might only be allowed access to the specified path prefix.
+	if b.pathPrefix != "" {
+		obj := <-s3Clnt.ListObjects(context.Background(), b.bucket, s3.ListObjectsOptions{Prefix: b.pathPrefix})
+		if obj.Err != nil {
+			typedErr := s3.ToErrorResponse(obj.Err)
+			if typedErr.Code != bucketNotFound {
+				return model.NewAppError("TestFileConnection", "api.file.test_connection.s3.list_objects.app_error", nil, obj.Err.Error(), http.StatusInternalServerError)
+			}
+			exists = false
```

Comment:
What happens if the bucket exists but the path prefix doesn't? Do we still get a `bucketNotFound` or else?
Should we attempt to recover from this by creating the path prefix (en empty directory I suppose)?

Reply:
That case was always handled. When we would loop, the loop would exit without running once. Now that we are seeking, so the default value of `obj` will come out of the channel. No err will be set. And everything will be hunky dory.

Reply:
So just to confirm I am getting this right: `ListObjects()` is not returning an error if we call it on an non-existent path under an existing bucket.

Reply:
Correct.